core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java

```diff
@@ -238,14 +238,14 @@ public static Block mapCast(
             MethodHandle keyProcessFunction,
             MethodHandle valueProcessFunction,
             Type targetType,
-            BlockPositionEqual keyEqual,
+            BlockPositionIsDistinctFrom keyDistinctOperator,
             BlockPositionHashCode keyHashCode,
             ConnectorSession session,
             Block fromMap)
     {
         checkState(targetType.getTypeParameters().size() == 2, "Expect two type parameters for targetType");
         Type toKeyType = targetType.getTypeParameters().get(0);
-        TypedSet resultKeys = createEqualityTypedSet(toKeyType, keyEqual, keyHashCode, fromMap.getPositionCount() / 2, "map-to-map cast");
+        TypedSet resultKeys = createDistinctTypedSet(toKeyType, keyDistinctOperator, keyHashCode, fromMap.getPositionCount() / 2, "map-to-map cast");
```

Comment:
You can somewhat test it with:
```
select cast(map(array['NaN', 'NaN '], array[1, 2]) as map<double, bigint>);
```

It will fail now. It used to not.

Reply:
Added a test.